### PR TITLE
Add monte carlo dropout

### PR DESCRIPTION
**Pull request by florian-huber**
First implementation to address #53 .
That might not yet be the most elegant way to implement things, but it could serve as a solid start.
- added `MS2DeepScoreMonteCarlo`
- added helper functions to compute the means and standard deviations

[edit] minor additional changes through merge with main (due to #66 )

### CODE REVIEW

File: ms2deepscore/models/SiameseModel.py [hunk 1]
```diff
@@ -105,34 +106,60 @@ def save(self, filename: Union[str, Path]):
...
+                       dropout_rate: float = 0.25,
+                       l1_reg: float = 1e-6,
+                       l2_reg: float = 1e-6,
+                       dropout_always_on: bool = False) -> keras.Model:
+        """Create base model for Siamaese network.
+
+        Parameters
+        ----------
+        input_dim : int
+            DESCRIPTION.
```
Comment (svenvanderburg) [inline ms2deepscore/models/SiameseModel.py:121]: I think this should be updated.

File: ms2deepscore/models/SiameseModel.py [hunk 2]
```diff
@@ -105,34 +106,60 @@ def save(self, filename: Union[str, Path]):
...
+        base_dims
```
Comment (svenvanderburg) [inline ms2deepscore/models/SiameseModel.py:122]: There is no `base_dims`, just `dims`

File: ms2deepscore/models/SiameseModel.py [hunk 3]
```diff
@@ -105,34 +106,60 @@ def save(self, filename: Union[str, Path]):
...
-                                                 name='dense'+str(i+1))(model_layer)
-            model_layer = keras.layers.BatchNormalization(name='normalization'+str(i+1))(model_layer)
-            model_layer = keras.layers.Dropout(dropout_rate, name='dropout'+str(i+1))(model_layer)
+                model_layer = Dense(dim, activation='relu', name='dense'+str(i+1))(model_layer)
+            model_layer = BatchNormalization(name='normalization'+str(i+1))(model_layer)
+            if dropout_always_on:
+                model_layer = Dropout(dropout_rate, name='dropout'+str(i+1))(model_layer,
+                                                                             training=True)
+            else:
+                model_layer = Dropout(dropout_rate, name='dropout'+str(i+1))(model_layer)
```
Comment (svenvanderburg) [inline ms2deepscore/models/SiameseModel.py:153]: Could be simplified to:
```suggestion

            model_layer = Dropout(dropout_rate, name='dropout'+str(i+1))(model_layer,
                                                                             training=dropout_always_on)
```

File: ms2deepscore/MS2DeepScoreMonteCarlo.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+from .typing import BinnedSpectrumType
+
+
+class MS2DeepScoreMonteCarlo(BaseSimilarity):
+    """Calculate MS2DeepScore ensemble similarity scores and STD between a reference
+    and a query using Monte-Carlo Dropout
+
+    Using a trained model, binned spectrums will be converted into spectrum
+    vectors using a deep neural network. The MS2DeepScoreMonteCarlo similarity is then
+    the mean of n_ensemble x n_ensemble cosine similarity score between two spectrum
```
Comment (svenvanderburg) [inline ms2deepscore/MS2DeepScoreMonteCarlo.py:17]: Is'nt it the man of n_ensemble cosine similarity scores? Instead of a matrix of n_ensemble x n_ensemble?

File: ms2deepscore/MS2DeepScoreMonteCarlo.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    def _create_input_vector(self, binned_spectrum: BinnedSpectrumType):
+        """Creates input vector for model.base based on binned peaks and intensities"""
+        X = np.zeros((1, self.input_vector_dim))
+
+        idx = np.array([int(x) for x in binned_spectrum.binned_peaks.keys()])
+        values = np.array(list(binned_spectrum.binned_peaks.values()))
+        X[0, idx] = values
+        return X
+
+    def _create_monte_carlo_encoder(self):
```
Comment (svenvanderburg) [inline ms2deepscore/MS2DeepScoreMonteCarlo.py:76]: I would stick to `base` instead of `encoder`

File: ms2deepscore/MS2DeepScoreMonteCarlo.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        Parameters
+        ----------
+        reference:
+            Reference spectrum.
+        query:
+            Query spectrum.
+
+        Returns
+        -------
+        ms2ds_similarity
```
Comment (svenvanderburg) [inline ms2deepscore/MS2DeepScoreMonteCarlo.py:107]: It returns the mean + std

File: ms2deepscore/MS2DeepScoreMonteCarlo.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+            Reference spectrum.
+        queries:
+            Query spectrum.
+        is_symmetric:
+            Set to True if references == queries to speed up calculation about 2x.
+            Uses the fact that in this case score[i, j] = score[j, i]. Default is False.
+
+        Returns
+        -------
+        ms2ds_similarity
```
Comment (svenvanderburg) [inline ms2deepscore/MS2DeepScoreMonteCarlo.py:134]: Returns matrix of mean similarity scores and matrix of std's

File: ms2deepscore/MS2DeepScoreMonteCarlo.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        progress_bar:
+            Set to True to monitor the embedding creating with a progress bar.
+            Default is False.
+        """
+        self.model = model
+        self.n_ensembles = n_ensembles
+        self.input_vector_dim = self.model.base.input_shape[1]  # TODO: later maybe also check against SpectrumBinner
+        self.output_vector_dim = self.model.base.output_shape[1]
+        self.progress_bar = progress_bar
+        self.partial_model = self._create_monte_carlo_encoder()
```
Comment (svenvanderburg) [inline ms2deepscore/MS2DeepScoreMonteCarlo.py:65]: Maybe be consistent in naming. I would suggest:
```python
self.monte_carlo_base = self._create_monte_carlo_base()
```

**Review by svenvanderburg — APPROVED** [review]
Looks good, I just added some small comments!

File: ms2deepscore/models/SiameseModel.py (hunk 3 above)
Comment (florian-huber) [reply ms2deepscore/models/SiameseModel.py:153]: That could cause trouble, because it would freeze the model in `training=False` for the default state. In that case dropout would stay off even during actual training.

File: ms2deepscore/models/SiameseModel.py (hunk 2 above)
Comment (florian-huber) [reply ms2deepscore/models/SiameseModel.py:122]: Done.

File: ms2deepscore/models/SiameseModel.py (hunk 1 above)
Comment (florian-huber) [reply ms2deepscore/models/SiameseModel.py]: Done

File: ms2deepscore/MS2DeepScoreMonteCarlo.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+        return X
+
+    def _create_monte_carlo_encoder(self):
+        """Rebuild base network with training=True"""
+        dims = []
+        for layer in self.model.base.layers:
+            if "dense" in layer.name:
+                dims.append(layer.units)
+            if "dropout" in layer.name:
+                dropout_rate = layer.rate
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply ms2deepscore/MS2DeepScoreMonteCarlo.py]: Thanks, good point. I now added a print statement if different dropout rate values are found.

File: ms2deepscore/MS2DeepScoreMonteCarlo.py (hunk 5 above)
Comment (florian-huber) [reply ms2deepscore/MS2DeepScoreMonteCarlo.py:76]: Done.

File: ms2deepscore/MS2DeepScoreMonteCarlo.py (hunk 8 above)
Comment (florian-huber) [reply ms2deepscore/MS2DeepScoreMonteCarlo.py]: Done

File: ms2deepscore/MS2DeepScoreMonteCarlo.py (hunk 6 above)
Comment (florian-huber) [reply ms2deepscore/MS2DeepScoreMonteCarlo.py]: done

File: ms2deepscore/MS2DeepScoreMonteCarlo.py (hunk 7 above)
Comment (florian-huber) [reply ms2deepscore/MS2DeepScoreMonteCarlo.py]: Added this to the documentation (and now also user can chose between median and mean). I recently thought that median would be better, because it is more robust against outliers, so I made median the default.

File: ms2deepscore/MS2DeepScoreMonteCarlo.py (hunk 4 above)
Comment (florian-huber) [reply ms2deepscore/MS2DeepScoreMonteCarlo.py:18]: It is indeed not totally intuitive... but `n_ensemble` means that `n_ensemble` embeddings are created per spectrum. And since we are interested in pair-wise scores, that means we can do `n_ensemble x n_ensemble` possible comparisons with that.

**Comment by florian-huber** [comment]
Thanks a lot for reviewing Sven!
I edited the code according to your comments.
I also changed the network architecture a bit to skip the dropout for the first layer (which is what I had done for my local screens anyway).
